File: beacon_monitoring_platform_interface/lib/src/model/monitoring_result.dart
// Copyright (c) 2020 Objectivity. All rights reserved.
// Use of this source code is governed by The MIT License (MIT) that can be
// found in the LICENSE file.

import 'dart:convert';

import 'package:flutter/foundation.dart';

import 'region.dart';

enum MonitoringEvent {
  didEnterRegion,
  didExitRegion,
  didDetermineState,
}

enum MonitoringState {
  inside,
  outside,
}

class MonitoringResult {
  static const _region = 'region';
  static const _type = 'type';
  static const _state = 'state';

  final Region region;
  final MonitoringEvent type;

  /// This value is not null when [type] is [MonitoringEvent.didDetermineState]
  final MonitoringState state;

  MonitoringResult({
    @required this.region,
    @required this.type,
    this.state,
  });

  MonitoringResult.fromJson(Map<String, dynamic> json)
      : region = Region.fromJson(json[_region]),
        type = MonitoringEventExtension.parse(json[_type]),
        state = MonitoringStateExtension.parse(json[_state]);

  Map<String, dynamic> toJson() => <String, dynamic>{
        _region: region.toJson(),
        _type: type.toJson(),
        _state: state?.toJson(),
      };

  @override
  String toString() => jsonEncode(toJson());
}

extension MonitoringEventExtension on MonitoringEvent {
  static parse(String value) {
    if (_StringExtension.isBlank(value)) return null;

    return MonitoringEvent.values.firstWhere((e) => describeEnum(e) == value);
  }

  String toJson() {
    return describeEnum(this);
  }
}

extension _StringExtension on String {
  static bool isBlank(String value) {
    return value == null || value.trim().isEmpty;
  }
}

extension MonitoringStateExtension on MonitoringState {
  static parse(String value) {
    if (_StringExtension.isBlank(value)) return null;

    return MonitoringState.values.firstWhere((e) => describeEnum(e) == value);
  }

  String toJson() {
    return describeEnum(this);
  }
}

File: beacon_monitoring/android/src/main/kotlin/uk/co/objectivity/beacon_monitoring/plugin/model/MonitoringResult.kt
// Copyright (c) 2020 Objectivity. All rights reserved.
// Use of this source code is governed by The MIT License (MIT) that can be
// found in the LICENSE file.

package uk.co.objectivity.beacon_monitoring.plugin.model

import android.os.Parcelable
import com.squareup.moshi.FromJson
import com.squareup.moshi.JsonClass
import com.squareup.moshi.Moshi
import com.squareup.moshi.ToJson
import kotlinx.android.parcel.Parcelize
import org.altbeacon.beacon.MonitorNotifier

fun MonitoringResult.toJson(): String = Moshi.Builder()
        .add(MonitoringEvent.Adapter())
        .add(MonitoringState.Adapter())
        .build()
        .adapter(MonitoringResult::class.java)
        .toJson(this)

enum class MonitoringEvent(val readableName: String) {
    didEnterRegion("Enter Region"),
    didExitRegion("Exit Region"),
    didDetermineState("Determine State");

    class Adapter {
        @FromJson
        fun fromJson(json: String): MonitoringEvent = valueOf(json)

        @ToJson
        fun toJson(value: MonitoringEvent): String = value.toString()
    }
}

enum class MonitoringState(private val value: Int) {
    inside(MonitorNotifier.INSIDE),
    outside(MonitorNotifier.OUTSIDE);

    companion object {
        fun fromInt(value: Int) = values().first { it.value == value }
    }

    class Adapter {
        @FromJson
        fun fromJson(json: String): MonitoringState = valueOf(json)

        @ToJson
        fun toJson(value: MonitoringState): String = value.toString()
    }
}

@Parcelize
@JsonClass(generateAdapter = true)
data class MonitoringResult(
        val region: RegionModel,
        val type: MonitoringEvent,
        val state: MonitoringState? = null
) : Parcelable